File: keymap_store.py
"""
keymap_store.py

Reads and writes the JSON keymap files described in docs/FILE_FORMAT.md.
This module is shared logic — it's deliberately kept free of any UI code
so the same read/write behavior can later be mirrored 1:1 in the C++
service (see cpp_service/keymap_store.h/.cpp for the native equivalent).
"""

import json
import os
import uuid
from datetime import datetime, timezone
from typing import Optional


SCHEMA_VERSION = 1

# In the real system-level build this would be a protected system path,
# e.g. /data/system/gaming_keyboard/keymaps/. For the prototype we use a
# local folder so it can run and be tested standalone.
DEFAULT_STORE_DIR = os.path.join(
    os.path.expanduser("~"), ".gaming_keyboard", "keymaps"
)


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _path_for(package_name: str, store_dir: str = DEFAULT_STORE_DIR) -> str:
    os.makedirs(store_dir, exist_ok=True)
    return os.path.join(store_dir, f"{package_name}.json")


def load_keymap(package_name: str, store_dir: str = DEFAULT_STORE_DIR) -> Optional[dict]:
    """Return the saved keymap dict for a package, or None if not mapped yet."""
    path = _path_for(package_name, store_dir)
    if not os.path.exists(path):
        return None
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def new_keymap(package_name: str, display_name: str, activation_key: str = "F5") -> dict:
    """Create a fresh in-memory keymap structure for a new mapping session."""
    now = _now_iso()
    return {
        "schema_version": SCHEMA_VERSION,
        "package_name": package_name,
        "display_name": display_name,
        "created_at": now,
        "updated_at": now,
        "activation_key": activation_key,
        "mappings": [],
    }


def add_mapping(keymap: dict, mapping_type: str, input_key: str,
                 position: tuple, action_label: str = "") -> dict:
    """Append a mapping entry (key or mouse) to the in-memory keymap.

    Returns the entry that was added, so the caller can track it for undo.
    """
    entry = {
        "id": uuid.uuid4().hex[:4],
        "type": mapping_type,
        "input_key": input_key,
        "action_label": action_label,
        "position": {"x": position[0], "y": position[1]},
    }
    keymap["mappings"].append(entry)
    return entry


def undo_last(keymap: dict) -> Optional[dict]:
    """Remove and return the most recently added mapping, if any."""
    if keymap["mappings"]:
        return keymap["mappings"].pop()
    return None


def save_keymap(keymap: dict, store_dir: str = DEFAULT_STORE_DIR) -> str:
    """Write the keymap to disk, overwriting any existing file for this
    package (per the spec: re-saving replaces the older mapping)."""
    keymap["updated_at"] = _now_iso()
    path = _path_for(keymap["package_name"], store_dir)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(keymap, f, indent=2)
    return path


def delete_keymap_file(package_name: str, store_dir: str = DEFAULT_STORE_DIR) -> bool:
    """Remove a game's keymap entirely (Settings list 'delete' action).
    Returns True if a file was actually removed."""
    path = _path_for(package_name, store_dir)
    if os.path.exists(path):
        os.remove(path)
        return True
    return False


def list_mapped_games(store_dir: str = DEFAULT_STORE_DIR) -> list:
    """Return [{'package_name':..., 'display_name':...}, ...] for the
    Settings > Physical Keyboard > Gaming Keyboard list."""
    if not os.path.isdir(store_dir):
        return []
    games = []
    for fname in os.listdir(store_dir):
        if fname.endswith(".json"):
            with open(os.path.join(store_dir, fname), "r", encoding="utf-8") as f:
                data = json.load(f)
                games.append({
                    "package_name": data["package_name"],
                    "display_name": data["display_name"],
                })
    return games

File: keymap_store.cpp
// keymap_store.cpp
//
// Implementation of KeymapStore. Uses nlohmann::json (header-only,
// https://github.com/nlohmann/json) for parsing — swap for whatever
// JSON library the Bliss OS build already depends on if there's a
// preferred one; the schema itself (docs/FILE_FORMAT.md) doesn't
// require this specific library.

#include "keymap_store.h"

#include <filesystem>
#include <fstream>
#include <chrono>
#include <random>
#include <sstream>
#include <iomanip>

#include <nlohmann/json.hpp>

namespace fs = std::filesystem;
using json = nlohmann::json;

namespace gaming_keyboard {

namespace {

std::string NowIso8601() {
    auto now = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm_utc{};
    gmtime_r(&t, &tm_utc);
    std::ostringstream oss;
    oss << std::put_time(&tm_utc, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}

std::string GenerateShortId() {
    static const char kChars[] = "0123456789abcdef";
    static std::mt19937 rng{std::random_device{}()};
    std::uniform_int_distribution<int> dist(0, 15);
    std::string id(4, '0');
    for (auto& c : id) c = kChars[dist(rng)];
    return id;
}

std::string MappingTypeToString(MappingType t) {
    return t == MappingType::kKey ? "key" : "mouse";
}

MappingType MappingTypeFromString(const std::string& s) {
    return s == "mouse" ? MappingType::kMouse : MappingType::kKey;
}

json MappingToJson(const MappingEntry& m) {
    return json{
        {"id", m.id},
        {"type", MappingTypeToString(m.type)},
        {"input_key", m.input_key},
        {"action_label", m.action_label},
        {"position", {{"x", m.position.x}, {"y", m.position.y}}},
    };
}

MappingEntry MappingFromJson(const json& j) {
    MappingEntry m;
    m.id = j.at("id").get<std::string>();
    m.type = MappingTypeFromString(j.at("type").get<std::string>());
    m.input_key = j.at("input_key").get<std::string>();
    m.action_label = j.value("action_label", "");
    m.position.x = j.at("position").at("x").get<int>();
    m.position.y = j.at("position").at("y").get<int>();
    return m;
}

json KeymapToJson(const Keymap& k) {
    json mappings = json::array();
    for (const auto& m : k.mappings) mappings.push_back(MappingToJson(m));

    return json{
        {"schema_version", k.schema_version},
        {"package_name", k.package_name},
        {"display_name", k.display_name},
        {"created_at", k.created_at},
        {"updated_at", k.updated_at},
        {"activation_key", k.activation_key},
        {"mappings", mappings},
    };
}

Keymap KeymapFromJson(const json& j) {
    Keymap k;
    k.schema_version = j.value("schema_version", 1);
    k.package_name = j.at("package_name").get<std::string>();
    k.display_name = j.value("display_name", k.package_name);
    k.created_at = j.value("created_at", "");
    k.updated_at = j.value("updated_at", "");
    k.activation_key = j.value("activation_key", "F5");
    if (j.contains("mappings")) {
        for (const auto& item : j.at("mappings")) {
            k.mappings.push_back(MappingFromJson(item));
        }
    }
    return k;
}

}  // namespace

KeymapStore::KeymapStore(std::string store_dir) : store_dir_(std::move(store_dir)) {
    fs::create_directories(store_dir_);
}

std::string KeymapStore::PathFor(const std::string& package_name) const {
    return (fs::path(store_dir_) / (package_name + ".json")).string();
}

std::optional<Keymap> KeymapStore::Load(const std::string& package_name) const {
    auto path = PathFor(package_name);
    if (!fs::exists(path)) return std::nullopt;

    std::ifstream in(path);
    if (!in) return std::nullopt;

    json j;
    in >> j;
    return KeymapFromJson(j);
}

Keymap KeymapStore::CreateNew(const std::string& package_name,
                               const std::string& display_name,
                               const std::string& activation_key) const {
    Keymap k;
    k.package_name = package_name;
    k.display_name = display_name;
    k.activation_key = activation_key;
    k.created_at = NowIso8601();
    k.updated_at = k.created_at;
    return k;
}

std::string KeymapStore::AddMapping(Keymap& keymap,
                                     MappingType type,
                                     const std::string& input_key,
                                     Position position,
                                     const std::string& action_label) {
    MappingEntry entry;
    entry.id = GenerateShortId();
    entry.type = type;
    entry.input_key = input_key;
    entry.action_label = action_label;
    entry.position = position;
    keymap.mappings.push_back(entry);
    return entry.id;
}

bool KeymapStore::UndoLast(Keymap& keymap) {
    if (keymap.mappings.empty()) return false;
    keymap.mappings.pop_back();
    return true;
}

std::string KeymapStore::Save(Keymap& keymap) const {
    keymap.updated_at = NowIso8601();
    auto path = PathFor(keymap.package_name);
    std::ofstream out(path);
    out << KeymapToJson(keymap).dump(2);
    return path;
}

bool KeymapStore::DeleteFile(const std::string& package_name) const {
    auto path = PathFor(package_name);
    if (!fs::exists(path)) return false;
    return fs::remove(path);
}

std::vector<GameListing> KeymapStore::ListMappedGames() const {
    std::vector<GameListing> result;
    if (!fs::exists(store_dir_)) return result;

    for (const auto& entry : fs::directory_iterator(store_dir_)) {
        if (entry.path().extension() != ".json") continue;

        std::ifstream in(entry.path());
        if (!in) continue;

        json j;
        in >> j;
        GameListing g;
        g.package_name = j.value("package_name", "");
        g.display_name = j.value("display_name", g.package_name);
        result.push_back(g);
    }
    return result;
}

}  // namespace gaming_keyboard
